### ISSUE TITLE
microphone-stream import bombing in Vite

### ISSUE THREAD

**Issue by reintroducing** [saebekassebil/microphone-stream#70]
## Describe the bug
I'm trying to port some code over from an older codebase that uses webpack to a new one that uses Vite. As such, I'm simply trying to import this module `import MicrophoneStream from 'microphone-stream';` and it bombs the page. Commenting the import out makes the page work again.

![Screen Shot 2022-02-25 at 2 18 49 PM](https://user-images.githubusercontent.com/259901/155804405-89110fa4-0ab3-48f3-a282-bfaf4da019cc.png)

I will not deny that not only am I new to working with voice input, but this is my first foray into Vite and it just so happens that within one day I ran up on this. I suspect that this is a result of the fact that Vite is working with ES Modules directly and require statements aren't supported at all, but ES Modules are the future of JS so I'm wondering if this is a Vite problem or a problem of this library (maybe there needs to be a module export in package.json added?). 

I suppose the goal of this is to open up the discussion around this issue and try and figure out where its stemming from.

#### Version number
`microphone-stream@version` 6.0.1

#### Is the issue in the browser/Node.js?
Browser (Chrome)

#### Details of the browser/Node.js version
Paste output of `npx envinfo --browsers` or `node -v`
Browsers:
    Chrome: 98.0.4758.109
    Firefox: 97.0
    Safari: 15.1

node v16.13.0

## Steps to reproduce
Go to https://stackblitz.com/edit/vitejs-vite-kar4vh?file=src%2FApp.jsx&terminal=dev and you will see the app working (this is Vite's basic start app for a React project). In App.jsx on line 4 I added (and commented out) the microphone-stream import. Uncomment that and watch the app blow up with the above error (this also happens locally, its not specific to StackBlitz, you can click the little cloud download button of the project to try it locally as well).

#### Observed behavior
The error above is displayed and the app bombs out.

#### Expected behavior
To import microphone-stream and use it just like any other module.

#### Screenshots
See above.

## Additional context
N/A

**Comment by nfriedly** [comment]
Oof, I'm sorry to hear that.

I'm only on my phone, but from a quick glance it looks like this is originating from the readable stream library that microphone stream depends on. In node.js, `global` is roughly equivalent to `window` in the browser, and I'm guessing readable stream is touching it for some reason.

Webpack is probably automatically rewrote that and apparently vite doesn't.

Try doing this somewhere before microphone-stream gets imported:

`window.global = window`

And let me know how it works. I'll try to look into a better solution when I have more time available, but hopefully that can get you unblocked.

**Comment by reintroducing** [comment]
I actually forked this repo and am working on some updates to support ESM (after thinking about this more I think that is the real problem here because I can see the compiled code is CJS only). If you'll accept it, I'll open up a PR with the changes needed to make this work. I'm fumbling through some of the TS stuff as we speak because I'm not a TS user, but I'm close to getting it done.

**Comment by nfriedly** [comment]
That's pretty cool, but I will warn you that it's unlikely to be as easy as it sounds. We just did that with the express-rate-limit package, and it was a big ordeal with a lot of testing and tweaking, and still a few patch releases after we shipped it.

**Comment by reintroducing** [comment]
I went ahead and opened the PR here: https://github.com/microphone-stream/microphone-stream/pull/71

It wasn't too bad, unless I'm missing something (highly likely)? :)

**Comment by reintroducing** [comment]
Ugh, i just tested the code generated by that PR and it does not fix the issue either :(  

I also tried adding `window.global = window;` but that unfortunately also did not fix the issue. I'm stumped.

**Comment by reintroducing** [comment]
Sorry, I don't mean to be spamming, but I'm just digging more and more and going down a bit of a rabbit hole. I believe you are correct in saying that it is caused by the global usage inside of readable-stream and there appears to be a PR to remove it (with lots of discussion in regards to bundlers and how they handle node globals) [here](https://github.com/nodejs/readable-stream/pull/435). 

It seems like development on readable-stream has kind of ground to a halt, so I suppose my last question is, is there any viable replacement for readable-stream that this package can use or am I basically SOL at this point?

**Comment by reintroducing** [comment]
quick follow up, putting `window.global = window` in my index.html file works but ONLY in development mode. the second you build the bundle, it breaks (even if you put it in the head, before the bundled js gets linked in the output html file). so, while it would unblock development, it will still not work in production unfortunately, so not a long term solution.

i've been doing a lot of reading on this over the past 24 hours and its basically Vite's (and webpack 5s) position to no longer include the polyfills for node code working in browser packages. webpack 5 does have a workaround by using a 3rd party polyfill but I can't seem to find one for Vite. not really sure where to go from here. :\

**Comment by nfriedly** [comment]
Do you know what the 3rd party polyfill is for readable-stream?

**Comment by nfriedly** [comment]
(It might make sense for this package to explicitly depend on it, rather than just letting the bundler do whatever it wants.)

**Comment by reintroducing** [comment]
@nfriedly it wouldn't necessarily be a polyfill for readable-stream, rather a polyfill for global. but to answer your question, its kind of hard to find answers around this, just a lot of back and forth between people who think bundlers like webpack 5 and vite (next gen bundlers basically) should continue to polyfill automatically and then those that want to push the ecosystem forward and not polyfill just because it used to be done in the past by webpack 4. One such discussion is [here](https://github.com/vitejs/vite/issues/4796) (with very strong opinions from both sides, and there are many more GH issues like this one that exist if you want to go down the rabbit hole like i did). [Here](https://github.com/vitejs/vite/discussions/5912) is another discussion (although brief and pointed). [Sindre's blog](https://blog.sindresorhus.com/webpack-5-headache-b6ac24973bf1) mentions the polyfill although not directly and as such I can't really find it (just apparently that it exists, possibly through the use of ProvidePlugin in webpack and even then I couldn't find the actual polyfill that would be applied through Provide Plugin.

**Comment by reintroducing** [comment]
following up again, because i found a workaround for the `global` issue (you can follow the discussion and last few comments [here](https://github.com/nodejs/readable-stream/pull/435)). Unfortunately I'm now stuck on another issue with buffer.from which is coming from `safe-buffer` which is a nested dependency inside of `readable-stream`. this is so much fun!

**Comment by reintroducing** [comment]
last follow up, i got it all working ([comment here](https://github.com/nodejs/readable-stream/pull/435#issuecomment-1053641748)). this has been one of the most not fun things i've done on a weekend in a long, long time :)

**Comment by nfriedly** [comment]
Wow, that does sound painful 😬. Nice job on finding a working solution, though.  

I'm still not really sure what the long-term answer should be :(

**Comment by reintroducing** [comment]
@nfriedly I think the long term solution for all package authors, if the industry heads towards not polyfilling these things in bundlers anymore, is going to be to not use any node specific packages in browser code. that is obviously easier said than done, though, as i don't even think many package authors are aware of this issue OR that their code is using node APIs because it has not been an issue for such a long time. 

I can see it from both sides. the people pushing the new tooling need this to become the new norm or it causes issues not only in their tooling but realistically is not right because node APIs weren't meant to be used in the browser. on the other hand, everyone just expects it to "just work" at this point because it has been for so long and these types of breaking changes have long term effects and will cause many issues for authors to have to account for (like this discussion we're having). if possible, it'd be great if there was a replacement that this package can use which does not use readable-stream, but I'm not familiar enough with the internals to know if thats possible.

**Comment by bryantwells** [comment]
thank you @reintroducing for documenting your solution. I am running up against the same issue – fingers crossed it doesn't break in the immediate future 🤞🏽

**Comment by seth-seikosoft** [comment]
I've just ran into this upgrading an old project from webpack to vite.  Hopefully your solution works for this as well!  You're awesome @reintroducing thank you!

**Comment by reintroducing** [comment]
@seth-seikosoft hopefully it works, but its unfortunate that a year and a half later this is still an issue :\

**Comment by nfriedly** [comment]
It looks like v4 of `readable-stream` depends on https://www.npmjs.com/package/buffer. I wonder if upgrading to `readable-streams@4` and dropping `buffer-from` in favor of the `buffer` package would solve the issue. Would someone like to try it out?

**Comment by seth-seikosoft** [comment]
> It looks like v4 of `readable-stream` depends on https://www.npmjs.com/package/buffer. I wonder if upgrading to `readable-streams@4` and dropping `buffer-from` in favor of the `buffer` package would solve the issue. Would someone like to try it out?

I've tried that and it looks to be working from the dev build but not the prod build.
I was following the answer #6 from here https://stackoverflow.com/questions/70714690/buffer-is-not-defined-in-react-vite 
```
<script>
    window.global = window;
</script>
<script type="module">
    import process from "process";
    import EventEmitter from "events";
    import {Buffer} from "buffer";
    window.Buffer = Buffer;
    window.process = process;
    window.EventEmitter = EventEmitter;
</script>
  ```
  
  I'm still trying to get the build to work but I currently have this error in my vite.config.ts
  `vite.config.ts:21:17 - error TS2769: No overload matches this call.`
  it doesn't like the `plugins: [GlobalPolyFill({ process: true, buffer: true })],` 
  
  vite.config.ts
  ```
import { defineConfig } from 'vite'
import react from '@vitejs/plugin-react'
import viteTsconfigPaths from 'vite-tsconfig-paths'
import GlobalPolyFill from '@esbuild-plugins/node-globals-polyfill'

export default defineConfig({
  plugins: [react(), viteTsconfigPaths()],
  server: {
    port: 5173,
  },
  resolve: {
    alias: {
      buff: 'rollup-plugin-node-polyfills/polyfills/buffer-es6',
    },
  },
  optimizeDeps: {
    esbuildOptions: {
      define: {
        global: 'globalThis',
      },
      plugins: [GlobalPolyFill({ process: true, buffer: true })],
    },
  },
  build: {
    outDir: 'build',
    rollupOptions: { external: [/@aws-sdk/] },
  },
})
```

**Comment by seth-seikosoft** [comment]
hey hey I think I might have gotten it.  At least when I run my esbuild command it it builds. 

vite.config.ts
```
import { defineConfig } from 'vite'
import react from '@vitejs/plugin-react'
import viteTsconfigPaths from 'vite-tsconfig-paths'
import GlobalPolyFill from '@esbuild-plugins/node-globals-polyfill'

export default defineConfig({
  plugins: [
    react(),
    viteTsconfigPaths(),
    GlobalPolyFill({ process: true, buffer: true }),
  ],
  server: {
    port: 5173,
  },
  resolve: {
    alias: {
      buff: 'rollup-plugin-node-polyfills/polyfills/buffer-es6',
    },
  },
  optimizeDeps: {
    esbuildOptions: {
      define: {
        global: 'globalThis',
      },
    },
  },
  build: {
    outDir: 'build',
    rollupOptions: { external: [/@aws-sdk/] },
  },
})
```
you need to stick the `GlobalPolyFill()` in the root level "plugins" and not in the nested one. 

then in your index.html you have in the `<head>` tags
```
<script type='module'>
    import { Buffer } from 'buffer'
    import EventEmitter from 'events'

    window.Buffer = Buffer
    window.EventEmitter = EventEmitter
</script>
```
